core/adminserver/admin_server.go

```diff
@@ -239,7 +237,14 @@ func (s *Server) CreateDomain(ctx context.Context, in *pb.CreateDomainRequest) (
 
 	// Initialize log with first map root.
 	if err := s.initialize(ctx, logTree, mapTree); err != nil {
-		return nil, fmt.Errorf("adminserver: initialize of log %v and map %v failed: %v",
+		// Delete log and map if initialization fails.
+		if _, delErr := s.logAdmin.DeleteTree(ctx, &tpb.DeleteTreeRequest{TreeId: logTree.TreeId}); delErr != nil {
+			return nil, status.Errorf(codes.Internal, "adminserver: CreateAndInitTree(map): %v, DeleteTree(%v): %v ", err, logTree.TreeId, delErr)
+		}
+		if _, delErr := s.mapAdmin.DeleteTree(ctx, &tpb.DeleteTreeRequest{TreeId: mapTree.TreeId}); delErr != nil {
```

Comment:
I wonder if in this case you ought to try deleting both the log and the map, and return a single error from which ever one fails?

i.e. if deleting the log fails, note that but also try deleting the map.

Reply:
Great idea.  Fewer if statements is always better.